### PR TITLE
components in blocks

### PR DESCRIPTION
**Pull request by iruzevic**

### CODE REVIEW

File: website/docs/basics/blocks-component-in-block.md [hunk 1]
```diff
@@ -11,16 +11,89 @@ Try to think of your development process in this way:
 
 - Each component must be standalone and ideally not depend on any other components (sometimes this is not possible but keep this to a minimum).
 - Components must not be aware of the layout they are used in.
-- No global styles. Instead, all styles should be contained to the block/component. For more details, check [this link](writing-styles).
+- No global styles. Instead, all styles should be contained to the block/component. For more details, check [this link](blocks-styles).
 - You should define all heading variations in a `heading` component. You can then use this component everywhere you have a heading instead of writing a new implementation. In practice: don't do `<h2>$heading</h2>` but render the `heading` component instead using the helpers defined in the [helpers section](helpers). The heading is just an example, but this applies to every component.
 - You can have multiple heading components in one block.
+- **JavaScript and PHP part of the block/component must be 99% te same this is why we wrote all those helpers.**
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:17]:
```suggestion
- **JavaScript and PHP part of the block/component must be 99% the same this is why we wrote all those helpers.**
```

File: website/docs/basics/blocks-component-in-block.md [hunk 2]
```diff
@@ -11,16 +11,89 @@ Try to think of your development process in this way:
...
+
+## Manifest setup
+
+In order to use our way or passing props you must understand a few things, you must use our helpers and you must set `components` key in the manifest.
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:21]:
```suggestion
In order to use our way of passing props, there are a few prerequisites you must follow. You must use our helpers and you must set `components` key in the manifest.
```

File: website/docs/basics/blocks-component-in-block.md [hunk 3]
```diff
@@ -11,16 +11,89 @@ Try to think of your development process in this way:
...
+
+As we described in the [block manifest](block-manifest) chapter each block/component can have `components` key. This key is used in the registration of the block attributes in the block editor and as a dependency tree for passing props in the childeren components.
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:23]:
```suggestion
As we described in the [block manifest](block-manifest) chapter each block/component can have a `components` key. This key is used in the registration of the block attributes in the block editor and as a dependency tree for passing props in the children components.
```

File: website/docs/basics/blocks-component-in-block.md [hunk 4]
```diff
@@ -11,16 +11,89 @@ Try to think of your development process in this way:
...
+**Default setup:**
+```json
+"components": {
+	"heading": "heading"
+}
+```
+
+Components key is an object that can have multiple items added. 
+
+> Key represents the name that is going to be used for this component in this block/component and the prefix used in tha attributes (more on this later).
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:34]:
```suggestion
> Key represents the name that is going to be used for this component in this block/component and the prefix used in the attributes (more on this later).
```

File: website/docs/basics/blocks-component-in-block.md [hunk 5]
```diff
@@ -11,16 +11,89 @@ Try to think of your development process in this way:
...
+```json
+"components": {
+	"heading": "heading",
+	"paragraph": "paragraph"
+}
+```
+
+This block/component uses two components, one is heading and the other is paragraphy and they are using the default names.
+
+**Multiple components setup with the same type:**
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:50]:
```suggestion
**Multiple components set up with the components of the same type:**
```

File: website/docs/basics/blocks-component-in-block.md [hunk 6]
```diff
@@ -11,16 +11,89 @@ Try to think of your development process in this way:
...
+```json
+"components": {
+	"heading": "heading",
+	"intro": "heading",
+	"paragraph": "paragraph"
+}
+```
+
+This block/component uses three components, one is heading and the other heading is renamed to intro and the last one is paragraphy.
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:59]:
```suggestion
This block/component uses three components, one is heading and the other heading is renamed to intro and the last one is paragraph.
```

File: website/docs/basics/blocks-component-in-block.md [hunk 7]
```diff
@@ -11,16 +11,89 @@ Try to think of your development process in this way:
...
+
+## Important helpers
+
+As we described earlier in order to use our setup you must use our helpers. Here are some of the helper you should get familiar with:
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:63]:
```suggestion
As we described earlier, in order to use our setup you must use our helpers. Here are some of the helpers you should get familiar with:
```

File: website/docs/basics/blocks-component-in-block.md [hunk 8]
```diff
@@ -11,16 +11,89 @@ Try to think of your development process in this way:
...
+* getOptions - [JS](https://github.com/infinum/eightshift-frontend-libs/blob/develop/scripts/editor/get-options.js) - [PHP](https://github.com/infinum/eightshift-libs/blob/develop/src/Helpers/Components.php)
+* getOption - [JS](https://github.com/infinum/eightshift-frontend-libs/blob/develop/scripts/editor/get-options.js) - [PHP](https://github.com/infinum/eightshift-libs/blob/develop/src/Helpers/Components.php)
+* getAttrKey - [JS](https://github.com/infinum/eightshift-frontend-libs/blob/develop/scripts/helpers/check-attr.js) - [PHP](https://github.com/infinum/eightshift-libs/blob/develop/src/Helpers/Components.php)
+* checkAttr - [JS](https://github.com/infinum/eightshift-frontend-libs/blob/develop/scripts/helpers/check-attr.js) - [PHP](https://github.com/infinum/eightshift-libs/blob/develop/src/Helpers/Components.php)
+* render - [PHP](https://github.com/infinum/eightshift-libs/blob/develop/src/Blocks/AbstractBlocks.php)
+
+For more details please check our helpers sections for [JavaScript](helpers-javascript) and [PHP](helpers-php).
+
+> Keep in mind that you can always ignore our setup and use the React native way of passing props to components by prop drilling.
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:74]:
```suggestion
> Keep in mind that you can always ignore our setup and use the React way of passing props to components by prop drilling.
```

React native way read like use `React native` and not `React` 😄

File: website/docs/basics/blocks-component-in-block.md [hunk 9]
```diff
@@ -11,16 +11,89 @@ Try to think of your development process in this way:
...
+
+
+## Setting up attributes
+
+> Each attribute must have the component name prefix in each attribute. This is one of the biggest requirement for our setup to work.
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:79]:
```suggestion
> Each attribute must have the component name prefix in each attribute. This is one of the biggest requirements in order for our setup to work.
```

File: website/docs/basics/blocks-component-in-block.md [hunk 10]
```diff
@@ -11,16 +11,89 @@ Try to think of your development process in this way:
...
+
+By defining components key in the block registration phase we are drilling down from block to the last component and setting up attribute names according to you setup. In order for everything to work we had to add/change prefixes to each component attribute in corelation where this attribute is used in the dependency tree. It is not so hard as it souns so let us show you some graphical examples:
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:81]:
```suggestion
By defining components key in the block registration phase we are drilling down from block to the last component and setting up attribute names according to your setup. In order for everything to work, we had to add/change prefixes to each component attribute in relation to where this attribute is used in the dependency tree. It is not so hard as it sounds so let us show you some graphical examples:
```

File: website/docs/basics/blocks-component-in-block.md [hunk 11]
```diff
@@ -41,8 +114,9 @@ and you want to use the heading component in the Jumbotron block, you can use it
 `src/Blocks/custom/jumbotron/manifest.json`
 
 ```json
+"blockName": "jumbotron",
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:117]: On line 112 there is a typo `component's` -> `component`

File: website/docs/basics/blocks-component-in-block.md [hunk 12]
```diff
@@ -57,147 +131,158 @@ In your block attributes object, you will now have these keys:
...
+"jumbotronHeadingLevel": 2,
+"jumbotronHeadingSize": "default",
 
 // Attributes from the block.
-"content": "test",
+"jumbotronContent": "test",
 ```
 
-As you can see in the example, you can merge component attributes in your block attributes. Keep in mind that, in the provided example, the components' attributes will be injected in the block attributes with the same name as defined in the components manifest.
+As you can see in the example, each block/component attribute must have prefix and we ara than stacking that prefixed in the final output.
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:142]:
```suggestion
As you can see in the example, each block/component attribute must have a prefix, and we are then stacking that prefixed in the final output.
```

File: website/docs/basics/blocks-component-in-block.md [hunk 13]
```diff
@@ -57,147 +131,158 @@ In your block attributes object, you will now have these keys:
...
 
-### I want to use one heading and paragraph component in my block.
+### I want to override one heading attribute in my block.
 
-You can do this by following the same principle as before. Here is an example:
+We build all our attributes by merging attributes objects in this order:
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:146]:
```suggestion
We build all our attributes by merging attribute's objects in this order:
```

File: website/docs/basics/blocks-component-in-block.md [hunk 14]
```diff
@@ -57,147 +131,158 @@ In your block attributes object, you will now have these keys:
...
-### I want to override a default attribute of a component in a block.
-
-We build all our attributes by merging attributes objects in this order:
+As you can see in the example, to overide tha component attribute you just set that attribute to a different value but you must respect the attribute prefix.
 
-* global attributes (global settings);
-* wrapper attributes;
-* component attributes; and
-* block attributes.
+### I want to use one heading and paragraph component in my block.
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:212]:
```suggestion
### I want to use one heading and a paragraph component in my block.
```

File: website/docs/basics/blocks-component-in-block.md [hunk 15]
```diff
@@ -248,15 +336,15 @@ in your block attributes object, you will now have these keys:
 ```json
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:336]:
```
in your block attribute's object, you will now have these keys:
```

File: website/docs/basics/blocks-component-in-block.md [hunk 16]
```diff
@@ -271,7 +359,7 @@ All our components come with a few common attributes that we recommend you also
...
 
-For example, you have a Jumbotron block and a heading that you want to populate if necessary. If you add content to the heading, everything looks good. But if you don't add it, you'll still see a placeholder for that heading in the block editor. That affects how you perceive it, and a user may think this block will look like that on the front end (with the placeholder text). So if you have a component that you will not populate on some block, switch the toggle to false, and you will hide it from the DOM. This is useful for content editors.
+For example, you have a Jumbotron block and a heading that you want to populate if necessary. If you add content to the heading, everything looks good. But if you don't add it, you'll still see a placeholder for that heading in the block editor. That affects how you perceive it, and a editor may think this block will look like that on the front end (with the placeholder text). So if you have a component that you will not populate on some block, switch the toggle to false, and you will hide it from the DOM. This is useful for content editors.
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:362]:
```suggestion
For example, you have a Jumbotron block and a heading that you want to populate if necessary. If you add content to the heading, everything looks good. But if you don't add it, you'll still see a placeholder for that heading in the block editor. This affects how you perceive it, and an editor may think this block will look like that on the front end (with the placeholder text). So if you have a component that you will not populate on some block, switch the toggle to false, and you will hide it from the DOM. This is useful for content editors.
```

File: website/docs/basics/blocks-component-in-block.md [hunk 17]
```diff
@@ -289,37 +377,36 @@ Exactly the same way as attributes.
...
-But in all places in your component where you use `setAttributes` you must use [camelize helper](https://github.com/infinum/eightshift-frontend-libs/blob/develop/scripts/helpers/camelize.js)
+For example, you have a `card` component that you will use in the `featured-posts` block.
 
-```js
+`src/Blocks/custom/featured-posts/featured-posts.php`
 
-import { camelize } from '@eightshift/frontend-libs/scripts/helpers';
+```php
+// Fetch and parse manifest using custom helper.
+$cardManifest = Components::getManifest(dirname(__DIR__, 2) . '/components/card');
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:394]: Where is this used in the example below?

Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:394]: Where is this going to be used in the example below?

File: website/docs/basics/blocks-component-in-block.md [hunk 18]
```diff
@@ -329,131 +416,130 @@ $cardManifest = Components::getManifest(dirname(__DIR__, 2) . '/components/card'
...
-Let's say you have a block that has a `Heading` component inside it.
+Let's say you have a block that has a `heading` component inside it.
 
-The `Heading` has 10 text sizes and 5 colors, but for that block, only 2 colors and 3 text sizes are allowed.
+The `heading` has 10 text sizes and 5 colors, but for that block, only 2 colors and 3 text sizes are allowed.
 To achieve this you must prepare your component to be able to use this feature:
 
 1. Your component options must be named the same name as their attribute value. In the example, you can see that the options key for `typographySize` is the same name as in the attributes.
 
+`src/Blocks/components/heading/heading.php`
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:439]:
```suggestion
`src/Blocks/components/heading/manifest.json`
```

File: website/docs/basics/blocks-component-in-block.md [hunk 19]
```diff
@@ -465,161 +551,148 @@ And this is it. You are now able to override the options from the parent block/c
 
 At one point in time, we agreed on naming standards for all component attributes. That way we made sure that you wouldn't get any collisions when using multiple components. So we said it is ok to spread all attributes to the component and let the component handle what it needs. Well, that approach is ok but it can bite you in the a.. at the point that you least expect.
 
-That is why we created this [props helper](helpers-javascript#props).
-
-As described in [this chapter](blocks-component-in-block#i-want-to-limit-which-options-are-shown-for-components-inside-a-blockcomponent) you must follow the attributes naming convention and use `components` key in the block/component manifest.
+That is why we created [props helper](helpers-javascript) that determins what attributes are passed to what component depending on the dependency tree.
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:554]:
```suggestion
That is why we created [props helper](helpers-javascript) that determines what attributes are passed to what component depending on the dependency tree.
```

File: website/docs/basics/blocks-component-in-block.md [hunk 20]
```diff
@@ -465,161 +551,148 @@ And this is it. You are now able to override the options from the parent block/c
...
+		Components::props('heading', $attributes)
+	);
+	?>
 </div>
 ```
 
-In PHP you just provide the results of the props helper. As you can see in the [props helper docs](helpers-javascript#props) you must provide the fourth parameter to be `true` to distinguish if this is a block or a component.
 
-Now, let's see how the `Component heading` looks like.
+Now, let's see how the `component heading` looks like.
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:625]:
```suggestion
Now, let's see how the component `heading` looks like.
```

File: website/docs/basics/blocks-component-in-block.md [hunk 21]
```diff
@@ -638,6 +711,35 @@ Components:
...
+```js
+<TypographyEditor
+	{...props('typography', attributes, {
+		typographyContent: item,
+	})}
+	disableEdit={true}
+/>
+```
+
+typographyContent attribute will follow the naming prefix when passed to a child but the disableEdit will remain the same name in all the childrens.
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:735]:
```suggestion
typographyContent attribute will follow the naming prefix when passed to a child but the disableEdit will remain the same name in all the children.
```

File: website/docs/basics/blocks-component-in-block.md [hunk 22]
```diff
@@ -638,6 +711,35 @@ Components:
...
+
+Here as a third parameter you can also pass all include attributes but they will not get prefixed:
```
Comment (dingo-d) [inline website/docs/basics/blocks-component-in-block.md:737]:
```suggestion
Here, as a third parameter, you can also pass all include attributes but they will not get prefixed:
```

**Review by dingo-d — CHANGES_REQUESTED** [review]
Left some comments. Stellar work. Now I actually kinda understand what all those helpers do 😂 

👍🏼

**Review by dingo-d — APPROVED** [review]

**Review by goranalkovic-infinum — APPROVED** [review]

File: website/docs/basics/blocks-component-in-block.md (hunk 11 above)
Comment (iruzevic) [reply website/docs/basics/blocks-component-in-block.md:117]: fixed

File: website/docs/basics/blocks-component-in-block.md [hunk 23]
```diff
@@ -216,9 +301,11 @@ If you have a heading component with these attributes:
 and you have a card component:
 
 `src/Blocks/components/card/manifest.json`
+
 ```json
+"componentName": "card",
 "attributes": {
-	"superCard": {
+	"cardUsage": {
```
[the remark this answers is not in the document]
Comment (iruzevic) [reply website/docs/basics/blocks-component-in-block.md:308]: this is just an example so no mater what it is as long as the reader understands it

File: website/docs/basics/blocks-component-in-block.md (hunk 17 above)
Comment (iruzevic) [reply website/docs/basics/blocks-component-in-block.md:394]: not used. I have removed it

**Pull request merged**